src/PointwiseFunctions/AnalyticSolutions/Elasticity/HalfSpaceMirror.cpp

```diff
@@ -0,0 +1,215 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/AnalyticSolutions/Elasticity/HalfSpaceMirror.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <gsl/gsl_sf_bessel.h>
+#include <pup.h>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "ErrorHandling/Exceptions.hpp"
+#include "NumericalAlgorithms/Integration/GslQuadAdaptive.hpp"
+#include "PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace Elasticity::Solutions {
+
+double displacement_r_integrand(const double k, const double r, const double z,
+                                const double beam_width,
+                                const double modulus_term_r) noexcept {
+  return gsl_sf_bessel_J1(k * r) * exp(-k * z - square(k * beam_width / 2.)) *
+         (modulus_term_r + k * z);
+}
+
+HalfSpaceMirror::HalfSpaceMirror(
+    const double beam_width, constitutive_relation_type constitutive_relation,
+    const size_t integration_intervals, const double absolute_tolerance,
+    const double relative_tolerance) noexcept
+    : beam_width_(beam_width),
+      constitutive_relation_(std::move(constitutive_relation)),
+      integration_intervals_(integration_intervals),
+      absolute_tolerance_(absolute_tolerance),
+      relative_tolerance_(relative_tolerance) {}
+
+tuples::TaggedTuple<Tags::Displacement<3>> HalfSpaceMirror::variables(
+    const tnsr::I<DataVector, 3>& x,
+    tmpl::list<Tags::Displacement<3>> /*meta*/) const noexcept {
+  const double shear_modulus = constitutive_relation_.shear_modulus();
+  const double lame_parameter = constitutive_relation_.lame_parameter();
+  auto result = make_with_value<tnsr::I<DataVector, 3>>(x, 0.);
+  const auto radius = sqrt(square(get<0>(x)) + square(get<1>(x)));
+
+  const integration::GslQuadAdaptive<
+      integration::GslIntegralType::UpperBoundaryInfinite>
+      integration{integration_intervals_};
+  const double lower_boundary = 0.;
+  const size_t num_points = get<0>(x).size();
+
+  const double prefactor = 0.25 / (shear_modulus * M_PI);
+  const double modulus_term_r = 1. - (lame_parameter + 2. * shear_modulus) /
+                                         (lame_parameter + shear_modulus);
+  const double modulus_term_z =
+      1. + shear_modulus / (lame_parameter + shear_modulus);
+  for (size_t i = 0; i < num_points; i++) {
+    const double z = get<2>(x)[i];
+    const double r = radius[i];
+    try {
+      if (r >= 100 * std::numeric_limits<double>::epsilon()) {
```

Comment:
Replace with `equal_within_roundoff(r, 0)` and include `Utilities/EqualWithinRoundoff.hpp` (same for the other occurrence in line 140 below)